blogs/2020/02/14/optimizing-ci.md

```diff
@@ -0,0 +1,117 @@
+
+## Room for Improvement
+
+So, what exactly were the common steps across all build jobs? Each build target has a job that follows a similar set of steps. At a very high level, each job must:
+
+1. Restore Node packages
+2. Lint the code
+3. Compile the code
+4. Run unit tests
+5. Run integration tests
+
+Our caching tasks were an obvious choice to speed up the “Restore Node packages” step. E.g. why run an expensive `npm install` step, when you could cache the results from a previous run, given that the `package-lock.json` file rarely changes? Since we’ve discussed caching packages previously, what makes this post interesting is how we applied caching to the other steps.
+
+Since linting and compilation are platform-independent, those steps could easily be run by a single build agent that would share its results with other platform-dependent agents. We created a Linux build agent whose sole responsibility was exactly this: restore packages, lint, and compile the code. All we had to do was to share the results with other agents.
+
+## Cache Everything
+
+In order to share cache results across build agents, we needed platform independent caches, which were initially unsupported by the caching tasks. We added a `platformIndependent` parameter to the task for creating and restoring platform-independent caches. For example:
+
+```yml
+- task: 1ESLighthouseEng.PipelineArtifactCaching.RestoreCacheV1.RestoreCache@1
+  inputs:
+    keyfile: keyfile
+    targetfolder: target
+    vstsFeed: $(ArtifactFeed)
+    platformIndependent: true
+```
+
+When caching node modules, it’s logical to use `package-lock.json` files as a cache key: when this file changes, we must invalidate the cache. When caching compilation outputs, the entire source code must act as a cache key. To simplify things, we decided to use the HEAD commit as the cache key—a new commit will inevitably create a new cache entry. This works fine for our purposes since a single build, despite running across build agents, always runs over a single commit.
+
+Another missing feature was the ability to create multiple caches per build job. We now found ourselves juggling two caches (node modules, compilation) with no way to address each cache individually. The caching tasks output an environment variable called `CacheRestored` that can be used to optimistically skip build tasks. This environment variable works great in builds that interact with a single cache, but not so great with multiple caches—leaving us to wonder which cache `CacheRestored` is referring to. Once again, we worked on a way to add an optional `alias` parameter to the task. For example:
```

Comment:
Again make it clear that this is not VS Code specific and that this is a pipelines extension

>Once again, we worked on a way to add an optional `alias` parameter to the task.